Modix.Bot/DiscordUserMessageTypeReader.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Discord;
+using Discord.Commands;
+using Microsoft.Extensions.DependencyInjection;
+using Modix.Data.Repositories;
+using Modix.Services.Core;
+
+namespace Modix
+{
+    public class DiscordUserMessageTypeReader : TypeReader
+    {
+        public override async Task<TypeReaderResult> ReadAsync(ICommandContext context, string input, IServiceProvider services)
+        {
+            if (ulong.TryParse(input, out var messageId))
+            {
+                var value = await FindMessageInUnknownChannelAsync(context, messageId, services);
+                var userMessage = value as IUserMessage;
+
+                if (userMessage is null)
+                    return TypeReaderResult.FromError(CommandError.ObjectNotFound, "Message not found.");
+
+                return TypeReaderResult.FromSuccess(new DiscordUserMessage(userMessage));
+            }
+
+            return TypeReaderResult.FromError(CommandError.ObjectNotFound, "Message not found.");
+        }
+
+        // TODO: extract to service layer after the PR that adds the DB check to the quote service version of this is merged.
+        private async Task<IMessage> FindMessageInUnknownChannelAsync(ICommandContext context, ulong messageId, IServiceProvider services)
+        {
+            var messageRepository = services.GetService<IMessageRepository>();
+            var guildMessage = await messageRepository.GetMessage(messageId);
+
+            if (guildMessage is { })
```

Comment:
...Am I just super out-of-the-loop? The hell is this syntax?

Reply:
It's equivalent to `!(guildMessage is null)`

Reply:
Is that like an anonymous type declaration?

Reply:
More like a recursive pattern matching thing.

In C# 8, you use property patterns like `someArray is { Length: 10 }`, which is equivalent to `!(someArray is null) && someArray.Length == 10`

So if you do `thing is { }`, it's the trivial/empty property pattern, which just ends up being the null check. Another way of writing it would be `thing is object`